Visualizes blushing on characters after using *blush or when they're drunk

kawoppi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Using the *blush emote makes your cheeks red for the next 5.2 seconds. The same effect is also visible while you are drunk (while the "drunk" mood event is added).
![group2](https://user-images.githubusercontent.com/94711066/164015033-8d014608-13d3-4e48-ac4d-96312818c075.PNG)
![plblush1](https://user-images.githubusercontent.com/94711066/164015179-c0c82cf6-ba4f-493a-a3b0-b4c5328c5bb3.PNG)
![plblush2](https://user-images.githubusercontent.com/94711066/164015191-258c764a-f589-4696-abd0-15b78cbc4306.PNG)
(sorry the plasmamen kept burning before i could get a good picture)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
It makes the characters a little bit more visually expressive. Also gives the HOP an easier time figuring out who's shitfaced while on the job.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
expansion: Blushing now makes your face go a little bit red
expansion: Being drunk flushes your cheeks
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

kawoppi committed:
Sprite on face test

Added blush sprite
Don't forget to finish all the things later on

kawoppi committed:
Improved sprite base color

kawoppi committed:
Blush activates using a trait

Blushing gives you the trait and and updates your body
Still need to find a way to remove it a moment later

kawoppi committed:
Moved trait

should probably be at the bottom of the mob ones?

kawoppi committed:
Tried to remove trait using addtimer

can't get it to work just yet

kawoppi committed:
IT FUCKING WORKS NOW

YEEEEEEAAAAAAAH

kawoppi committed:
Changed timing and added comment code

the science just isnt there yet

kawoppi committed:
Brings the code up to speed with the quirk branch

Used defines
Used only one timer
Buried hopes and dreams
Made the timer reset when blushing again
Made the sprite opaque
Made sprite appear behind eyes (gee thanks, moths)

Onule commented:
Could this also be added to drunk characters please?

kawoppi committed:
Copied conflicting file

kawoppi committed:
Added the blush sprite back in

wesoda25 commented:
Cute

kawoppi committed:
Merge remote-tracking branch 'upstream/master' into dev/blush

please work the way i think it does there are 400 files oh fuck

kawoppi committed:
Apply blush visual when drunk

It gets applied when the "drunk" mood event signal happens
Still need to find a way to get rid of the magic number

kawoppi committed:
Used trait for drunken blush effect

mood events!
so extendable!
so cool!

kawoppi committed:
Added full path for parameter

I think this should fix the error

kawoppi committed:
i forgot a symbol

that's probably why there's an error

thgvr commented:
Now THIS is content.

MrMelbert commented on code/__DEFINES/traits.dm:
```diff
@@ -359,6 +359,10 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_DONT_WRITE_MEMORY "dont_write_memory"
 /// This mob can be painted with the spraycan
 #define TRAIT_SPRAY_PAINTABLE "spray_paintable"
+/// This person is blushing
+#define TRAIT_BLUSHING "blushing"
+/// This person is blushing due to being drunk
+#define TRAIT_DRUNK_BLUSHING "drunk_blushing"
```
Traits can actually be assigned and handled from different sources, meaning you don't need both of these.
You only need one trait: `TRAIT_BLUSHING`

kawoppi replied:
Thanks for the review! I've added EMOTE_TRAIT and DRUNK_TRAIT defines to use as trait sources

MrMelbert commented on code/datums/mood_events/drink_events.dm:
```diff
@@ -2,6 +2,16 @@
 	mood_change = 3
 	description = "Everything just feels better after a drink or two."
 
+/datum/mood_event/drunk/add_effects(param)
+	// Display blush visual
+	ADD_TRAIT(owner, TRAIT_DRUNK_BLUSHING, SPECIES_TRAIT)
+	owner.update_body()
+
+/datum/mood_event/drunk/remove_effects()
+	// Stop displaying blush visual
+	REMOVE_TRAIT(owner, TRAIT_DRUNK_BLUSHING, SPECIES_TRAIT)
+	owner.update_body()
```
The third argument for add and remove trait is the source, and usually it's a descriptor or key related to what added it. 

In this case you're using `SPECIES_TRAIT` which doesn't seem accurate - replace it with something like "DRUNK_TRAIT" or a similar define

MrMelbert commented on code/modules/mob/living/emote.dm:
```diff
@@ -4,10 +4,34 @@
+	. = ..()
+	if(. && isliving(user))
+		var/mob/living/living_user = user
+		ADD_TRAIT(living_user, TRAIT_BLUSHING, SPECIES_TRAIT)
```
Same here - instead of applying via "species_trait", use a key like "emote" or something

bluetshirtguy commented:
the skyrat pipeline...

Ryll-Ryll commented:
15 seconds feels a little long, no?

kawoppi committed:
Used trait sources instead of two traits

kawoppi committed:
Changed blush duration to 5.2 seconds

This seems to be about as long as the message above your head sticks around for. Aligns with CHAT_MESSAGE_SPAWN_TIME + CHAT_MESSAGE_LIFESPAN but that might not be the  timing used for emotes since adding CHAT_MESSAGE_EOL_FADE makes it last too long.

kawoppi commented:
Good thing you brought that up! I set the duration when I was still trying to do other things with it and I didn't think to take another look at it. I've changed it to 5.2 seconds, which seems to match up with how long the text above your head is visible. Although I haven't been able to find the exact duration of it in the code.

kawoppi commented:
they had something like this already!? maybe this codebase should be a skyrat downstream

Mothblocks commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -581,6 +581,14 @@ GLOBAL_LIST_EMPTY(features_by_species)
 				standing += eye_overlay
 
+		// blush
+		#define BLUSH_COLOR "#DE5D83"
```
Put this in colors instead as like `COLOR_BLUSH_PINK` or similar.

Mothblocks commented on code/modules/mob/living/emote.dm:
```diff
@@ -4,10 +4,34 @@
+			// The existing timer restarts if it's already running
+			blush_timer = addtimer(CALLBACK(living_emote, /datum/emote/living/blush.proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
+
+/datum/emote/living/blush/proc/end_blush(mob/living/living_user)
```
Might want to check that `living_user` is not `QDELETED`

Mothblocks commented on code/modules/mob/living/emote.dm:
```diff
@@ -4,10 +4,34 @@
+		var/list/key_emotes = GLOB.emote_list["blush"]
+		for(var/datum/emote/living/blush/living_emote in key_emotes)
+			// The existing timer restarts if it's already running
+			blush_timer = addtimer(CALLBACK(living_emote, /datum/emote/living/blush.proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
```
```suggestion
			blush_timer = addtimer(CALLBACK(living_emote, .proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
```

Mothblocks commented on code/__DEFINES/traits.dm:
```diff
@@ -628,6 +630,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAUMA_TRAIT "trauma"
 #define FLIGHTPOTION_TRAIT "flightpotion"
+#define EMOTE_TRAIT "emote"
+#define DRUNK_TRAIT "drunk"
```
I guess I won't require it but in the future I always prefer to use `[type]` for this one off stuff.

kawoppi replied:
Thanks for the review! I'm not too familiar with this, but I've replaced the trait sources with "[type]" and it looks to be functioning just as before. Hopefully that lines up with how it should be used.

kawoppi committed:
Moved color define to colors

kawoppi committed:
Shortens callback proc

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

kawoppi committed:
Merge remote-tracking branch 'origin/dev/blush' into dev/blush

kawoppi committed:
Added QDELETED check

kawoppi committed:
Replaced trait source defines with [type]

that's neat
